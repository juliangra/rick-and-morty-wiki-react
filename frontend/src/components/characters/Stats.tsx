import { Center, RingProgress } from '@mantine/core'
import { MAX_RATING } from 'src/constants/rating'

type StatsProps = {
  rating: number
  testId?: string
}

const Stats: React.FC<StatsProps> = ({ rating, testId }) => {
  // The rating is a number between 0 and 100, but our implementation of rating is between 0 and 5
  const calculatedRating = (rating * 100) / MAX_RATING

  return (
    <RingProgress
      size={80}
      roundCaps
      thickness={8}
      sections={[{ value: calculatedRating, color: 'indigo' }]}
      label={<Center>{rating.toFixed(1)}</Center>}
      data-testid={testId}
      data-cy={testId}
    />
  )
}

export default Stats
