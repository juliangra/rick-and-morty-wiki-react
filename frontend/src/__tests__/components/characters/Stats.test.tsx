import { screen } from '@testing-library/react'
import Stats from 'src/components/characters/Stats'
import { createWithProviders, renderWithProviders } from 'src/utils/tests/render'

describe('Stats', () => {
  it('matches snapshot', () => {
    const component = createWithProviders(<Stats rating={1} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correct label', async () => {
    const rating = 5
    await renderWithProviders(<Stats rating={rating} testId={'stats'} />)
    expect(screen.getByTestId('stats')).toHaveTextContent(String(rating))
  })

  it('renders label with 1 decimal place', async () => {
    const rating = 3.33333
    await renderWithProviders(<Stats rating={rating} testId={'stats'} />)
    expect(screen.getByTestId('stats')).toHaveTextContent(String(rating.toFixed(1)))
  })
})
