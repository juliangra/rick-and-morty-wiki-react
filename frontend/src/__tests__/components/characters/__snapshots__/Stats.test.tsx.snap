// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats matches snapshot 1`] = `
<div
  className="mantine-RingProgress-root mantine-12sbrde"
  style={
    Object {
      "height": 80,
      "width": 80,
    }
  }
>
  <svg
    height={80}
    style={
      Object {
        "transform": "rotate(-90deg)",
      }
    }
    width={80}
  >
    <circle
      cx={40}
      cy={40}
      fill="none"
      r={28}
      stroke="#f1f3f5"
      strokeDasharray="140.74335088082273, 35.18583772020568"
      strokeDashoffset={0}
      strokeLinecap="butt"
      strokeWidth={8}
      transform="scale(1, -1) translate(0, -80)"
    />
    <circle
      cx={40}
      cy={40}
      fill="none"
      r={28}
      stroke="#4c6ef5"
      strokeDasharray="35.18583772020568, 140.74335088082273"
      strokeDashoffset={175.92918860102841}
      strokeLinecap="round"
      strokeWidth={8}
      transform={null}
    />
  </svg>
  <div
    className="mantine-37msp1 mantine-RingProgress-label"
    style={
      Object {
        "left": 16,
        "right": 16,
      }
    }
  >
    <div
      className="mantine-8od8ev"
    >
      1.0
    </div>
  </div>
</div>
`;
